tools/bowtie2.wdl

```diff
@@ -0,0 +1,410 @@
+#@ except: ContainerValue, CommentWhitespace
+
+version 1.1
+
+import "../data_structures/read_group.wdl"
+
+### Bowtie2 accepts several function parameters.
+### The first term is a function type. Available function types are:
+### - C - constant
+### - L - linear
+### - S - square-root
+### - G - natural logarithm
+### The constant and coefficient types may be negative and/or floating point numbers.
+###
+### An example input JSON entry might look like:
+### ```
+### {
+###     "interval_seed_substrings": {
+###         "function_type": "S",
+###         "constant": 1,
+###         "coefficient": 0.50,
+###     },
+### }
+### ```
+### See the function documentation in bowtie2
+### (https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#setting-function-options)
+struct Bowtie2Function {
+    String function_type
+    Float constant
+    Float coefficient
+}
+
+task build {
+    meta {
+        description: "Builds a Bowtie2 index from a FASTA file"
+        outputs: {
+            index_files: "Array of Bowtie2 index files",
+            bowtie_db_tar_gz: "Bowtie2 index files in tarred, gzipped format",
+        }
+    }
+
+    parameter_meta {
+        reference: "Reference genome in FASTA format (optionally gzipped)"
+        prefix: "Prefix for the Bowtie2 index files"
+        ncpu: "Number of threads to use"
+    }
+
+    input {
+        File reference
+        String prefix
+        Int ncpu = 4
+    }
+
+    String base = basename(reference, ".gz")
+
+    command <<<
+        set -euo pipefail
+
+        gunzip -c ~{reference} > ~{base} \
+           || ln -sf ~{reference} ~{base}
+
+        bowtie2-build --threads=~{ncpu} ~{base} ~{prefix}
+
+        tar -czf ~{prefix}.tar.gz ~{prefix}.*
+    >>>
+
+    output {
+        Array[File] index_files = glob("*.bt2")
+        File bowtie_db_tar_gz = prefix + ".tar.gz"
+    }
+
+    runtime {
+        cpu: ncpu
+        memory: "20 GB"
+        container: "ghcr.io/stjudecloud/bowtie2:2.5.4"
+        maxRetries: 1
+    }
+}
+
+task align {
+    meta {
+        description: "Aligns reads to a reference genome using Bowtie2"
+        outputs: {
+            aligned_bam: "Aligned reads in BAM format",
+            unpaired_unaligned: "Unpaired reads that didn't align",
+            unpaired_aligned: "Unpaired reads that aligned at least once",
+            paired_discordant: "Pairs that didn't align concordantly",
+            paired_concordant: "Pairs that aligned concordantly at least once",
+            metrics: "Metrics file",
```

Comment:
What format are all these files? From looking at the extensions I can't tell, other than for `unpaired_unaligned` which appears to be FASTQ. Regardless, the format should be documented here.

Should `unpaired_aligned` be empty assuming a well formed input file?